arch/arm/src/sama5/sam_flexcom_spi.c

```diff
@@ -0,0 +1,2116 @@
+  struct sam_flex_spidev_s *flex_spi = flex_spi_dev(flex_spics);
+  uint32_t regval;
+
+  /* Are we selecting or de-selecting the device? */
+
+  spiinfo("selected=%d\n", selected);
+  if (selected)
+    {
+      spiinfo("cs=%d\n", flex_spics->cs);
+
+      /* Before writing the TDR, the PCS field in the SPI_MR register must be
+       * set in order to select a slave.
+       */
+
+      regval  = flex_spi_getreg(flex_spi, SAM_FLEXCOM_SPI_MR_OFFSET);
+      regval &= ~FLEXCOM_SPI_MR_PCS_MASK;
+      regval |= (flex_spi_cs2pcs(flex_spics) << FLEXCOM_SPI_MR_PCS_SHIFT);
+      flex_spi_putreg(flex_spi, regval, SAM_FLEXCOM_SPI_MR_OFFSET);
+    }
+
+  /* Perform any board-specific chip select operations. PIO chip select
+   * pins may be programmed by the board specific logic in one of two
+   * different ways.  First, the pins may be programmed as SPI peripherals.
+   * In that case, the pins are completely controlled by the SPI driver.
+   * The sam_spi[0|1]select methods still needs to be provided, but they
+   * may be only stubs.
+   *
+   * An alternative way to program the PIO chip select pins is as normal
+   * PIO outputs.  In that case, the automatic control of the CS pins is
+   * bypassed and this function must provide control of the chip select.
+   * NOTE:  In this case, the PIO output pin does *not* have to be the
+   * same as the NPCS pin normal associated with the chip select number.
+   */
+
+  flex_spi->select(devid, selected);
+}
+
+/****************************************************************************
+ * Name: flex_spi_setfrequency
+ *
+ * Description:
+ *   Set the SPI frequency.
+ *
+ * Input Parameters:
+ *   dev  - Device-specific state data
+ *   freq - The SPI frequency requested
+ *
+ * Returned Value:
+ *   Returns the actual frequency selected
+ *
+ ****************************************************************************/
+
+static uint32_t flex_spi_setfrequency(struct spi_dev_s *dev, uint32_t freq)
+{
+  struct sam_flex_spics_s *flex_spics = (struct sam_flex_spics_s *)dev;
+  struct sam_flex_spidev_s *flex_spi = flex_spi_dev(flex_spics);
+  uint32_t actual;
+  uint32_t scbr;
+  uint32_t dlybs;
+  uint32_t dlybct;
+  uint32_t regval;
+  unsigned int offset;
+
+  spiinfo("cs=%d frequency=%" PRId32 "\n", flex_spics->cs, freq);
+
+  /* Check if the requested frequency is the same as the frequency
+   * selection
+   */
+
+  if (flex_spics->frequency == freq)
+    {
+      /* We are already at this frequency.  Return the actual. */
+
+      return flex_spics->actual;
+    }
+
+  /* Configure SPI to a frequency as close as possible to the requested
+   * frequency.
+   *
+   *   SPCK frequency = SPI_CLK / SCBR, or SCBR = SPI_CLK / frequency
+   */
+
+  scbr = SAM_FLEXCOM_SPI_CLOCK / freq;
+
+  if (scbr < 8)
+    {
+      scbr = 8;
+    }
+  else if (scbr > 254)
+    {
+      scbr = 254;
+    }
+
+  scbr = (scbr + 1) & ~1;
+
+  /* Save the new scbr value */
+
+  offset = (unsigned int)g_csroffset[flex_spics->cs];
```

Comment:
I do not think that cast is needed here

Reply:
I see now that sam_spi.c has g_csroffset as unit8_t as the offsets are within the range of a byte, but for flexcom they are not so I changed it to uint16_t here. "offset" is an unsigned int which is 32 bits on this processor so I think the cast may technically be unnecessary as no information is likely to be lost without the case, but isn't it good practice to do this?